File: src/leetCode/java/Solution332.java
package leetCode.java;

import java.util.*;

//Hierholzer's Algorithm O(ElogE) O(E), 時間複雜度為演算法本身O(E)乘上PriorityQueue所需O(logE)
class Solution332 {
    public List<String> findItinerary(List<List<String>> tickets) {
        //1 edges
        //PriorityQueue因為return the itinerary that has the smallest lexical order
        Map<String, PriorityQueue<String>> edges = new HashMap<>();//<from, <to1, ...>>
        for(List<String> ticket : tickets){
            if(!edges.containsKey(ticket.get(0)))
                edges.put(ticket.get(0), new PriorityQueue<String>());

            edges.get(ticket.get(0)).offer(ticket.get(1));
        }

        //2 DFS
        List<String> res = new ArrayList<>();
        dfs("JFK", edges, res);

        return res;
    }

    public void dfs(String node, Map<String, PriorityQueue<String>> edges, List<String> res){
        while(edges.containsKey(node) && !edges.get(node).isEmpty()){
            String child = edges.get(node).poll();//O(logn)
            dfs(child, edges, res);
        }
        res.add(0, node);//沒有路走了, 開始逆序入棧
    }
}
/*
A -> B --> C -> D
	 ^
     |
	 v
 	 E

初次走到D後回溯並加入 => BCD
回到B繼續DFS, 加入 => BEB BCD
回到A => A BEB BCD
*/

File: src/leetCode/c++/Solution332.cpp
//Hierholzer's Algorithm O(ElogE) O(E), 時間複雜度為演算法本身O(E)乘上PriorityQueue所需O(logE)
class Solution332 {
public:
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        //return the itinerary lexical order, 透過pq, 文字從小到大
        unordered_map<string, priority_queue<string, vector<string>, greater<string>>> edges;//<起點, <目的地1, ...>>
        for(auto ticket : tickets)
            edges[ticket[0]].push(ticket[1]);

        vector<string> res;
        dfs("JFK", edges, res);
        reverse(res.begin(), res.end());//矯正逆序入列
        return res;
    }

    void dfs(string node, unordered_map<string, priority_queue<string, vector<string>, greater<string>>>& edges, vector<string>& res){
        while(edges.count(node) && !edges[node].empty()){//回到此層後若還有會繼續while, 遍歷所有可能
            string child = edges[node].top();
            edges[node].pop();
            dfs(child, edges, res);
        }

        res.push_back(node);
    }
};